app/tests/heartbeat_controller_desktop_test.cc

```diff
@@ -0,0 +1,345 @@
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent1);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(300));
+
+  // Log a heartbeat for UserAgent2 on day2
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent2);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(300));
+
+  // Log a heartbeat for UserAgent1 on day3
+  app_common::RegisterLibrariesFromUserAgent(kCustomUserAgent1);
+  controller_.LogHeartbeat();
+  std::this_thread::sleep_for(std::chrono::milliseconds(300));
+
+  // Read from the storage class to verify
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage_.ReadTo(read_heartbeats);
+  ASSERT_TRUE(read_ok);
+  EXPECT_EQ(read_heartbeats.last_logged_date, day3);
+  ASSERT_EQ(read_heartbeats.heartbeats.size(), 2);
+  // UserAgent1 was logged on day1 and day3
+  ASSERT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1].size(), 2);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1][0], day1);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent1][1], day3);
+  // UserAgent2 was logged on day2
+  ASSERT_EQ(read_heartbeats.heartbeats[kCustomUserAgent2].size(), 1);
+  EXPECT_EQ(read_heartbeats.heartbeats[kCustomUserAgent2][0], day2);
+}
+
+TEST_F(HeartbeatControllerDesktopTest, LogMoreThan30DaysRemovesOldEntries) {
+  {
+    // InSequence guarantees that all of the expected calls occur in order.
+    testing::InSequence seq;
+    for (int month = 1; month <= 3; month++) {
+      for (int day = 1; day <= 30; day++) {
+        // YYYY-MM-DD\0
+        char date[11];
+        snprintf(date, sizeof(date), "2000-%02d-%02d", month, day);
+        std::string date_string(date);
+        EXPECT_CALL(mock_date_provider_, GetDate())
+            .WillOnce(Return(date_string));
+      }
+    }
+  }
+  for (int i = 1; i <= 90; i++) {
+    controller_.LogHeartbeat();
+  }
+  // Since LogHeartbeat is done asynchronously, wait a bit before verifying that
+  // the log succeeded.
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // Read from the storage class to verify
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage_.ReadTo(read_heartbeats);
+  ASSERT_TRUE(read_ok);
+  EXPECT_EQ(read_heartbeats.last_logged_date, "2000-03-30");
+  ASSERT_EQ(read_heartbeats.heartbeats.size(), 1);
+  // Even though heartbeat logging is asynchronous, it happens in the order that
+  // it is scheduled So the heartbeats should be logged in the correct order.
+  for (auto const& entry : read_heartbeats.heartbeats) {
+    std::vector<std::string> dates = entry.second;
+    ASSERT_EQ(dates.size(), 30);
+    EXPECT_EQ(dates[0], "2000-03-01");
+    EXPECT_EQ(dates[29], "2000-03-30");
+  }
+}
+
+TEST_F(HeartbeatControllerDesktopTest, DestroyControllerWhileWorkIsScheduled) {
+  HeartbeatController* destructible_controller =
+      new HeartbeatController(kAppId, logger_, mock_date_provider_);
+  {
```

Comment:
This bracket and indentation is not necessary

Reply:
Removed indendation